docs: dynamic datasource, model, repository, and controller (#5630)
strongloop/loopback-next: docs/site/Repositories.md

## Creating Repositories at Runtime

Repositories can be created at runtime using the `defineCrudRepositoryClass`
helper function from the `@loopback/rest-crud` package. It creates
`DefaultCrudRepository`-based repository classes by default.

```ts
const BookRepository = defineCrudRepositoryClass<
  Book,
  typeof Book.prototype.id,
  BookRelations
>(BookModel);
```

In case you want to use a non-`DefaultCrudRepository` repository class or you
want to create a custom repository, use the `defineRepositoryClass()` helper
function instead. Pass a second parameter to this function as the base class for
the new repository.

There are two options for doing this:

#### 1. Using a base repository class

Create a base repository with your custom implementation, and then specify this
repository as the base class.

```ts
class MyRepoBase<
  E extends Entity,
  IdType,
  Relations extends object
> extends DefaultCrudRepository<E, IdType, Relations> {
  // Custom implementation
}

const BookRepositoryClass = defineRepositoryClass<
  typeof BookModel,
  MyRepoBase<BookModel, typeof BookModel.prototype.id, BookRelations>
>(BookModel, MyRepoBase);
```

#### 2. Using a Repository mixin

> Create a repository mixin with your customization as shown in the "

Reviewer: Suggestion: since it will be a link, i think we don't need the double quotes.

[Defining A Repository Mixin Class Factory Function](https://loopback.io/doc/en/lb4/migration-models-mixins.html#defining-a-repository-mixin-class-factory-function)
example, apply the mixin on the base repository class (e.g.
`DefaultCrudRepository`) then specify this combined repository as the base class
to be used.

```ts
const BookRepositoryClass = defineRepositoryClass<
  typeof BookModel,
  DefaultCrudRepository<
    BookModel,
    typeof BookModel.prototype.id,
    BookRelations
  > &
    FindByTitle<BookModel>
>(BookModel, FindByTitleRepositoryMixin(DefaultCrudRepository));
```

Dependency injection has to be configured for the datasource as shown below.

```ts
inject(`datasources.${dsName.name}`)(BookRepository, undefined, 0);
const repoBinding = app.repository(BookRepository);
```

> Note, the `app.repository()` method will be available only on `RepositoryMixin`
> apps

Reviewer 2: Per the discussion in https://github.com/strongloop/loopback-next/pull/5630/files#r433340133, I am proposing the following:

```suggestion
Note, the `app.repository()` method is available only application classes with `RepositoryMixin` applied.
```

Please repeat this comment for Models and Repositories too.

> apps

Reviewer: Application class?

Author: Nah, it's not available on the `Application` class, it has to be composed with `RepositoryMixin`.

Reviewer: Right. I was thinking whether there's a better description than "RepositoryMix apps". I think it can be read it as Application class with RepositoryMixin?
